text/0000-javascript-module-api.md

```diff
@@ -0,0 +1,1131 @@
+utility functions should be a named export. That means you can never have more
+than one class per module, and _that_ means, inherently, more `import`
+statements than a system where multiple classes can live in one module.
+
+Additionally, in cases where there is not a clear "primary" class, this can feel
+a little awkward:
+
+```js
+import Route from "@ember/routing/route";
+import Router from "@ember/routing/router";
+```
+
+One commonly proposed alternative is to say that classes become named exports,
+and default exports are not used at all. The above example would become:
+
+```js
+import { Route, Router } from "@ember/routing";
+```
+
+In this case, classes are distinguished by being capitalized, rather than by
+being a default export.
+
+There is one major change coming to JavaScript and Ember that, your author
+believes, deals a fatal blow to this approach: decorators.
+
+If you're unfamiliar with decorators, see [Addy Osmani's great
+overview][decorators]. Decorators provide a mechanism for adding declarative
+annotations to classes, methods, properties and functions.
+
+[decorators]: https://medium.com/google-developers/exploring-es7-decorators-76ecb65fb841#.y9umddvsl
+
+For example, Robert Jackson has an [experimental library for using decorators to
+annotate computed properties in a class][ember-computed-decorators]. Something
+like this will probably make its way into Ember in the future:
+
+```js
+import EmberObject, { computed } from "@ember/object";
+
+export default class Cat extends EmberObject {
+  @computed("hairLength")
+  isDomesticShortHair(hairLength) {
+    return hairLength < 3;
+  }
+}
+```
+
+[ember-computed-decorators]: https://github.com/rwjblue/ember-computed-decorators
+
+Most decorators are tightly coupled to a particular class because they configure
+some aspect of behavior that is only relevant to that class. If every decorator
+and every class share a namespace, it is hard to identify which go with each
+other.
+
+```js
+import { Router, Route, resource, model, location, inject, queryParam } from "@ember/routing";
+```
+
+Can you tell me which of these decorators goes with which class?
+
+And this import is getting so long, you'd probably be tempted to break it up
+into multiple lines _anyway_, so it's not clear that it's actually a win over
+separate imports.
+
+Contrast this with the same thing expressed using the rules in
+this RFC:
+
+```js
+import Router, { resource, location } from "@ember/routing/router";
+import Route, { model, inject, queryParam } from "@ember/routing/route";
+```
+
+Here, the decorators are clearly tied to their class. And it's far nicer from a
+refactoring perspective: if you delete a class from a file, you then delete a
+single line from your imports.
+
+Contrast that with making fiddly edits to a long list of named exports unrelated
+to each other.
+
+# Unresolved Questions
+
+### Intimate APIs
+
+How much do we want to provide module API for so-called "intimate
```

Comment:
@kellyselden I'm not opposed to doing that, but opting in all private API by default seems likely to make the problem worse.

I think I'd prefer to say that we will surface intimate APIs on a case by case basis. For the case of `expandProperties`, this seems like a feature with very small scope, so would be a good candidate for an RFC to make it public.